internal/teststep/config.go

```diff
@@ -0,0 +1,310 @@
+
+type ConfigurationRequest struct {
+	Directory string
+	Raw       string
+}
+
+func Configuration(req ConfigurationRequest) (configuration, error) {
+	var populatedConfig []string
+	var config configuration
+
+	if req.Directory != "" {
+		populatedConfig = append(populatedConfig, fmt.Sprintf("%q", "directory"))
+
+		config = configuration{
+			directory: req.Directory,
+		}
+	}
+
+	if req.Raw != "" {
+		populatedConfig = append(populatedConfig, fmt.Sprintf("%q", "raw"))
+
+		config = configuration{
+			raw: req.Raw,
+		}
+	}
+
+	if len(populatedConfig) > 1 {
+		return configuration{}, fmt.Errorf(
+			"both %s are populated, only one configuration option is allowed",
+			strings.Join(populatedConfig, " and "),
+		)
+	}
+
+	return config, nil
+}
+
+func (c configuration) HasConfiguration() bool {
+	if c.directory != "" {
+		return true
+	}
+
+	if c.raw != "" {
+		return true
+	}
+
+	return false
+}
+
+func (c configuration) HasConfigurationFiles() bool {
+	return c.directory != ""
+}
+
+// HasProviderBlock returns true if the Config has declared a provider
+// configuration block, e.g. provider "examplecloud" {...}
+func (c configuration) HasProviderBlock(ctx context.Context) (bool, error) {
+	switch {
+	case c.hasRaw(ctx):
+		return providerConfigBlockRegex.MatchString(c.raw), nil
+	case c.hasDirectory(ctx):
+		pwd, err := os.Getwd()
+
+		if err != nil {
+			return false, err
+		}
+
+		configDirectory := filepath.Join(pwd, c.directory)
+
+		contains, err := filesContains(configDirectory, providerConfigBlockRegex)
+
+		if err != nil {
+			return false, err
+		}
+
+		return contains, nil
+	}
+
+	return false, nil
+}
+
+// HasTerraformBlock returns true if the Config has declared a terraform
+// configuration block, e.g. terraform {...}
+func (c configuration) HasTerraformBlock(ctx context.Context) (bool, error) {
+	switch {
+	case c.hasRaw(ctx):
+		return terraformConfigBlockRegex.MatchString(c.raw), nil
+	case c.hasDirectory(ctx):
+		pwd, err := os.Getwd()
+
+		if err != nil {
+			return false, err
+		}
+
+		configDirectory := filepath.Join(pwd, c.directory)
+
+		contains, err := filesContains(configDirectory, terraformConfigBlockRegex)
+
+		if err != nil {
+			return false, err
+		}
+
+		return contains, nil
+	}
+
+	return false, nil
+}
+
+func (c configuration) Write(ctx context.Context, dest string) error {
+	switch {
+	case c.directory != "":
+		err := c.writeDirectory(ctx, dest)
+
+		if err != nil {
+			return err
+		}
+	case c.raw != "":
+		err := c.writeRaw(ctx, dest)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copyFiles(path string, dstPath string) error {
+	infos, err := os.ReadDir(path)
+
+	if err != nil {
+		return err
+	}
+
+	for _, info := range infos {
+		srcPath := filepath.Join(path, info.Name())
+
+		if info.IsDir() {
+			continue
+		} else {
+			err = copyFile(srcPath, dstPath)
+
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func copyFile(path string, dstPath string) error {
```

Comment:
[`internal/plugintest`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-testing/internal/plugintest) has some helpers for this sorta thing -- if it makes more sense, maybe that functionality can split into its own internal file handling package to help prevent duplicating logic, but totally optional of course.

Reply:
I've left this as is for now if that's ok? Once these changes are merged I'll add an issue to consider refactoring.